File: football-api/app/services/recurrence.py
import asyncio
import secrets
from datetime import date, datetime, timezone, timedelta

import structlog
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.repositories.group_repo import GroupRepository
from app.db.repositories.match_repo import MatchRepository
from app.db.session import get_session_factory
from app.models.match import Attendance, AttendanceStatus, Match, MatchStatus
from app.services.push import send_push

_MONTHS_PT = ["jan","fev","mar","abr","mai","jun","jul","ago","set","out","nov","dez"]

def _fmt_date(d) -> str:
    return f"{d.day} de {_MONTHS_PT[d.month - 1]}"

logger = structlog.get_logger()


def _generate_hash() -> str:
    return secrets.token_urlsafe(8)[:10]


async def run_recurrence(session: AsyncSession) -> int:
    """
    Para cada grupo com recorrência ativa:
    - Se já há uma partida aberta, não faz nada.
    - Se a última partida já passou (match_date < hoje), cria a próxima
      com data = última + 7 dias, herdando configurações e convidados (status pending).

    Retorna o número de partidas criadas.
    """
    g_repo = GroupRepository(session)
    m_repo = MatchRepository(session)

    groups = await g_repo.get_groups_with_recurrence()
    created = 0
    today = datetime.now(timezone(timedelta(hours=-3))).date()

    for group in groups:
        if await m_repo.has_open_match(group.id):
            logger.info("recurrence_skipped_open_match", group_id=str(group.id))
            continue

        last_match = await m_repo.get_last_match(group.id)
        if not last_match:
            logger.info("recurrence_skipped_no_match", group_id=str(group.id))
            continue

        if last_match.match_date > today:
            logger.info("recurrence_skipped_future_match", group_id=str(group.id))
            continue

        if last_match.match_date == today and last_match.status != MatchStatus.CLOSED:
            logger.info("recurrence_skipped_today_not_closed", group_id=str(group.id))
            continue

        next_date = last_match.match_date + timedelta(days=7)

        # Garantia de hash único
        for _ in range(5):
            hash_ = _generate_hash()
            if not await m_repo.get_by_hash(hash_):
                break

        new_match = Match(
            group_id=group.id,
            match_date=next_date,
            start_time=last_match.start_time,
            end_time=last_match.end_time,
            location=last_match.location,
            address=last_match.address,
            court_type=last_match.court_type,
            players_per_team=last_match.players_per_team,
            max_players=last_match.max_players,
            notes=last_match.notes,
            hash=hash_,
            status=MatchStatus.OPEN,
            created_by_id=None,
        )
        session.add(new_match)
        await session.flush()

        player_ids = await m_repo.get_attendance_player_ids(last_match.id)
        for player_id in player_ids:
            session.add(Attendance(
                match_id=new_match.id,
                player_id=player_id,
                status=AttendanceStatus.PENDING,
            ))

        await session.flush()

        match_url = f"https://rachao.app/match/{hash_}"
        await asyncio.gather(*[
            send_push(
                session, pid,
                title=f"⚽ Novo rachão — {group.name}",
                body=f"Partida em {_fmt_date(next_date)}. Confirme sua presença!",
                url=match_url,
            )
            for pid in player_ids
        ], return_exceptions=True)

        created += 1
        logger.info(
            "recurrence_match_created",
            group_id=str(group.id),
            match_date=str(next_date),
            inherited_players=len(player_ids),
        )

    return created


async def _send_in_progress_pushes(session, m_repo: MatchRepository, candidates: list) -> None:
    """Envia push 'Bola rolando' para os jogadores confirmados dos candidatos a IN_PROGRESS."""
    for m in candidates:
        confirmed_ids = await m_repo.get_confirmed_player_ids(m.id)
        await asyncio.gather(*[
            send_push(
                session, pid,
                title=f"⚽ Bola rolando! — {m.group.name}",
                body="A partida de hoje já começou! 🎉",
                url=f"https://rachao.app/match/{m.hash}",
            )
            for pid in confirmed_ids
        ], return_exceptions=True)


async def run_status_sync_job() -> None:
    """
    Job horário — fecha partidas com data passada e transiciona para IN_PROGRESS.
    Não cria novos rachões (responsabilidade exclusiva do job das 07h).
    """
    session_factory = get_session_factory()
    async with session_factory() as session:
        try:
            m_repo = MatchRepository(session)
            in_progress_candidates = await m_repo.get_in_progress_candidates()
            closed = await m_repo.close_past_matches()
            await session.commit()
            if closed:
                logger.info("status_sync_auto_closed", matches_closed=closed)
            await _send_in_progress_pushes(session, m_repo, in_progress_candidates)
        except Exception:
            await session.rollback()
            logger.exception("status_sync_job_failed")


async def run_recurrence_job() -> None:
    """Job das 07h — fecha partidas passadas, cria próximos rachões e envia notificações."""
    session_factory = get_session_factory()
    async with session_factory() as session:
        try:
            m_repo = MatchRepository(session)
            in_progress_candidates = await m_repo.get_in_progress_candidates()
            closed = await m_repo.close_past_matches()
            if closed:
                logger.info("recurrence_auto_closed", matches_closed=closed)
            count = await run_recurrence(session)
            await session.commit()
            logger.info("recurrence_job_done", matches_created=count)
            await _send_in_progress_pushes(session, m_repo, in_progress_candidates)
        except Exception:
            await session.rollback()
            logger.exception("recurrence_job_failed")

File: football-api-go/internal/services/recurrence.go
package services

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"log/slog"
	"time"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5/pgxpool"

	"github.com/thiagotn/football-manager/football-api-go/internal/db"
)

var monthsPT = [...]string{"jan", "fev", "mar", "abr", "mai", "jun", "jul", "ago", "set", "out", "nov", "dez"}

func fmtDatePT(d time.Time) string {
	return fmt.Sprintf("%d de %s", d.Day(), monthsPT[d.Month()-1])
}

func generateMatchHash() (string, error) {
	b := make([]byte, 6)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b)[:10], nil
}

// matchVotingClosed reports whether the voting window for m has already passed.
func matchVotingClosed(m *db.Match) bool {
	matchDate, err := time.Parse("2006-01-02", m.MatchDate)
	if err != nil {
		return false
	}
	baseTimeStr := m.StartTime
	if m.EndTime != nil {
		baseTimeStr = *m.EndTime
	}
	baseTime, err := time.Parse("15:04:05", baseTimeStr)
	if err != nil {
		return false
	}
	// Reconstruct as BRT (UTC-3) wall-clock moment in UTC
	matchRef := time.Date(
		matchDate.Year(), matchDate.Month(), matchDate.Day(),
		baseTime.Hour(), baseTime.Minute(), baseTime.Second(), 0,
		time.UTC,
	).Add(3 * time.Hour) // shift BRT → UTC

	votingOpens := matchRef.Add(time.Duration(m.VoteOpenDelayMinutes) * time.Minute)
	votingCloses := votingOpens.Add(time.Duration(m.VoteDurationHours) * time.Hour)
	return time.Now().UTC().After(votingCloses)
}

// sendPushToGroup sends a push notification to all players in a group (best-effort stub).
func sendPushToGroup(ctx context.Context, pool *pgxpool.Pool, groupID uuid.UUID, matchURL, title, body string) {
	playerIDs, err := db.GetGroupMemberPlayerIDs(ctx, pool, groupID)
	if err != nil {
		return
	}
	for _, pid := range playerIDs {
		subs, err := db.GetPushSubscriptionsForPlayer(ctx, pool, pid)
		if err != nil {
			continue
		}
		for i := range subs {
			_ = SendPushToSubscription(
				subs[i].Endpoint, subs[i].P256dh, subs[i].Auth,
				"", "",
				PushNotification{Title: title, Body: body, URL: matchURL},
			)
		}
	}
}

// RunRecurrence creates the next recurring match for each eligible group.
// A group is eligible when: recurrence_enabled=true, no open match exists,
// the last match date has passed, and its voting window is closed.
// Returns the number of matches created.
func RunRecurrence(ctx context.Context, pool *pgxpool.Pool) (int, error) {
	groups, err := db.GetGroupsWithRecurrence(ctx, pool)
	if err != nil {
		return 0, fmt.Errorf("recurrence: list groups: %w", err)
	}

	brtNow := time.Now().UTC().Add(-3 * time.Hour)
	today := time.Date(brtNow.Year(), brtNow.Month(), brtNow.Day(), 0, 0, 0, 0, time.UTC)

	created := 0
	for _, group := range groups {
		hasOpen, err := db.HasOpenMatch(ctx, pool, group.ID)
		if err != nil {
			slog.Error("recurrence: check open match", "group_id", group.ID, "error", err)
			continue
		}
		if hasOpen {
			continue
		}

		lastMatch, err := db.GetLastMatch(ctx, pool, group.ID)
		if err != nil || lastMatch == nil {
			continue
		}

		lastDate, err := time.Parse("2006-01-02", lastMatch.MatchDate)
		if err != nil {
			continue
		}

		if lastDate.After(today) {
			continue
		}
		if lastDate.Equal(today) && lastMatch.Status != "closed" {
			continue
		}
		if !matchVotingClosed(lastMatch) {
			continue
		}

		nextDate := lastDate.Add(7 * 24 * time.Hour)

		var hash string
		for range 5 {
			h, err := generateMatchHash()
			if err != nil {
				break
			}
			existing, _ := db.GetMatchByHash(ctx, pool, h)
			if existing == nil {
				hash = h
				break
			}
		}
		if hash == "" {
			slog.Error("recurrence: unique hash not found", "group_id", group.ID)
			continue
		}

		num, err := db.NextMatchNumber(ctx, pool, group.ID)
		if err != nil {
			slog.Error("recurrence: next match number", "group_id", group.ID, "error", err)
			continue
		}

		newMatch, err := db.CreateMatchForRecurrence(ctx, pool, db.CreateMatchRecurrenceParams{
			GroupID:              group.ID,
			Hash:                 hash,
			Number:               num,
			MatchDate:            nextDate.Format("2006-01-02"),
			StartTime:            lastMatch.StartTime,
			EndTime:              lastMatch.EndTime,
			Location:             lastMatch.Location,
			Address:              lastMatch.Address,
			CourtType:            lastMatch.CourtType,
			PlayersPerTeam:       lastMatch.PlayersPerTeam,
			MaxPlayers:           lastMatch.MaxPlayers,
			Notes:                lastMatch.Notes,
			VoteOpenDelayMinutes: group.VoteOpenDelayMinutes,
			VoteDurationHours:    group.VoteDurationHours,
		})
		if err != nil {
			slog.Error("recurrence: create match", "group_id", group.ID, "error", err)
			continue
		}

		playerIDs, err := db.GetGroupMemberPlayerIDs(ctx, pool, group.ID)
		if err != nil {
			slog.Error("recurrence: get member ids", "group_id", group.ID, "error", err)
			continue
		}
		if err := db.CreateAttendances(ctx, pool, newMatch.ID, playerIDs); err != nil {
			slog.Error("recurrence: create attendances", "group_id", group.ID, "error", err)
			continue
		}

		matchURL := "https://rachao.app/match/" + hash
		sendPushToGroup(ctx, pool, group.ID, matchURL,
			"⚽ Novo rachão — "+group.Name,
			"Partida em "+fmtDatePT(nextDate)+". Confirme sua presença!",
		)

		created++
		slog.Info("recurrence: match created",
			"group_id", group.ID,
			"match_date", nextDate.Format("2006-01-02"),
			"players", len(playerIDs),
		)
	}
	return created, nil
}

// RunStatusSyncJob closes past matches and transitions today's matches to in_progress.
func RunStatusSyncJob(ctx context.Context, pool *pgxpool.Pool) error {
	candidates, err := db.GetInProgressCandidates(ctx, pool)
	if err != nil {
		return fmt.Errorf("status sync: get candidates: %w", err)
	}

	closed, err := db.ClosePastMatches(ctx, pool)
	if err != nil {
		return fmt.Errorf("status sync: close past: %w", err)
	}
	if closed > 0 {
		slog.Info("status_sync: closed past matches", "count", closed)
	}

	for _, c := range candidates {
		ids := []uuid.UUID{c.ID}
		if err := db.TransitionToInProgress(ctx, pool, ids); err != nil {
			slog.Error("status_sync: transition", "match_id", c.ID, "error", err)
			continue
		}
		matchURL := "https://rachao.app/match/" + c.Hash
		sendPushToGroup(ctx, pool, c.GroupID, matchURL,
			"⚽ Bola rolando! — "+c.GroupName,
			"A partida de hoje já começou! 🎉",
		)
	}
	return nil
}
